rust/chains/tw_polkadot/src/extrinsic.rs

```diff
@@ -0,0 +1,554 @@
+const IDENTITY_ADD_AUTHORIZATION: &str = "Identity.add_authorization";
+
+type CallIndex = (u8, u8);
+type CallIndicesTable = HashMap<&'static str, CallIndex>;
+
+macro_rules! call_indices {
+    ($($chain:expr => { $($name:expr => ($($value:expr),*) $(,)?)* } $(,)? )*) => {
+        [
+            $((
+                $chain, std::collections::HashMap::from_iter(
+                    [$(($name, ($($value as u8),+))),+]
+                )
+            )),+
+        ].into_iter().collect()
+    }
+}
+
+lazy_static! {
+    static ref CALL_INDICES_BY_NETWORK: HashMap<NetworkId, CallIndicesTable> = call_indices! {
+        NetworkId::POLKADOT => {
+            BALANCE_TRANSFER          => (0x05, 0x00),
+            STAKING_BOND              => (0x07, 0x00),
+            STAKING_BOND_EXTRA        => (0x07, 0x01),
+            STAKING_CHILL             => (0x07, 0x06),
+            STAKING_NOMINATE          => (0x07, 0x05),
+            STAKING_REBOND            => (0x07, 0x13),
+            STAKING_UNBOND            => (0x07, 0x02),
+            STAKING_WITHDRAW_UNBONDED => (0x07, 0x03),
+            UTILITY_BATCH             => (0x1a, 0x02),
+        },
+        NetworkId::KUSAMA => {
+            BALANCE_TRANSFER          => (0x04, 0x00),
+            STAKING_BOND              => (0x06, 0x00),
+            STAKING_BOND_EXTRA        => (0x06, 0x01),
+            STAKING_CHILL             => (0x06, 0x06),
+            STAKING_NOMINATE          => (0x06, 0x05),
+            STAKING_REBOND            => (0x06, 0x13),
+            STAKING_UNBOND            => (0x06, 0x02),
+            STAKING_WITHDRAW_UNBONDED => (0x06, 0x03),
+            UTILITY_BATCH             => (0x18, 0x02),
+        }
+    };
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum EncodeError {
+    InvalidNetworkId,
+    MissingCallIndicesTable,
+    InvalidCallIndex,
+    InvalidAddress,
+    InvalidValue,
+}
+
+type EncodeResult<T> = Result<T, EncodeError>;
+
+impl ToScale for SS58Address {
+    fn to_scale_into(&self, out: &mut Vec<u8>) {
+        Raw(self.key_bytes()).to_scale_into(out)
+    }
+}
+
+#[derive(Debug, Clone, Default)]
+pub struct ExtrinsicEncoder {
+    encoded: Vec<u8>,
+}
+
+impl ExtrinsicEncoder {
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    fn from_encoded(encoded: Vec<u8>) -> Self {
+        Self { encoded }
+    }
+
+    pub fn add(self, t: impl ToScale) -> Self {
+        let mut encoded = self.encoded;
+        t.to_scale_into(&mut encoded);
+        Self::from_encoded(encoded)
+    }
+
+    pub fn add_mut(&mut self, t: impl ToScale) -> &mut Self {
+        t.to_scale_into(&mut self.encoded);
+        self
+    }
+
+    pub fn finalize(self) -> Vec<u8> {
+        self.encoded
+    }
+}
+
+// `Extrinsic` is (for now) just a lightweight wrapper over the actual protobuf object.
+// In the future, we will refine the latter to let the caller specify arbitrary extrinsics.
```

Comment:
I personally prefer to separate business logic (low-level blockchain implementation) and TrustWallet Protobuf interface. Please consider adding a `src/extrinsic_builder.rs` or `src/extrinsic_protobuf.rs` that will try to convert Protobuf messages to Extrinsics.